driver/src/orderbook/batched_auction_data_reader.rs

```diff
@@ -0,0 +1,341 @@
+    /// The total number of orders per user.
+    user_order_counts: HashMap<H160, usize>,
+}
+
+/// Data for the next call to the smart contract's `getEncodedUsersPaginated`
+/// function (which should be named `getEncodedOrdersPaginated`).
+pub struct Pagination {
+    /// The user of the last received order or H160::zero when no order has been
+    /// received.
+    pub previous_page_user: H160,
+    /// The number of received orders for `previous_page_user`.
+    pub previous_page_user_offset: usize,
+}
+
+impl BatchedAuctionDataReader {
+    /// Create a new BatchedAuctionDataReader.
+    pub fn new(index: U256) -> BatchedAuctionDataReader {
+        BatchedAuctionDataReader {
+            account_state: AccountState::default(),
+            orders: Vec::new(),
+            pagination: Pagination {
+                previous_page_user: H160::zero(),
+                previous_page_user_offset: 0,
+            },
+            index,
+            user_order_counts: HashMap::new(),
+        }
+    }
+
+    /// The pagination data used when reading data from the smart contract in
+    /// batches.
+    pub fn pagination(&self) -> &Pagination {
+        &self.pagination
+    }
+
+    /// Signal that no more data will be read and return the result consuming
+    /// self.
+    pub fn get_auction_data(self) -> (AccountState, Vec<Order>) {
+        (self.account_state, self.orders)
+    }
+
+    /// Applies one batch of data.
+    ///
+    /// A batch can come from `getEncodedUsersPaginated` or `getEncodedOrders`.
+    /// In the latter case there is only one batch.
+    ///
+    /// Returns the number of orders in the data, before filtering for `index`.
+    ///
+    /// Panics if length of `packed_auction_bytes` is not a multiple of
+    /// `AUCTION_ELEMENT_WIDTH`.
+    pub fn apply_batch(&mut self, packed_auction_bytes: &[u8]) -> usize {
+        let auction_elements = self.parse_auction_elements(packed_auction_bytes);
+        let number_of_orders = auction_elements.len();
+        if number_of_orders == 0 {
+            return 0;
+        }
+
+        let last_order_user = auction_elements
+            .last()
+            .expect("there are no orders")
+            .order
+            .account_id;
+        self.apply_auction_data(auction_elements);
+        self.pagination.previous_page_user = last_order_user;
+        self.pagination.previous_page_user_offset = *self
+            .user_order_counts
+            .get(&last_order_user)
+            .expect("user has order but no order count");
+        number_of_orders
+    }
+
+    fn apply_auction_data(&mut self, auction_elements: Vec<StableXAuctionElement>) {
+        self.apply_auction_elements_to_account_state(auction_elements.iter());
+        // Workaround for borrow checker that would complain that `extend`
+        // borrows self as mutable while at the same time the filter
+        // closure borrows self as immutable because of using `self.index`.
+        let index = self.index;
```

Comment:
Do we still need this workaround now that we are not borrowing self mutably in the same method anymore?

Reply:
It is still needed.

Reply:
I see. It says why in the comment.